File: frontend/src/api/school.ts
import { apiClient } from "./client";
import type {
  ScoreLine,
  SchoolDetail,
  SchoolListItem,
  SchoolScoreLineParams,
  SchoolSearchParams,
  SchoolSearchResponse
} from "../types/school";

// 兼容后端历史字段：旧接口可能返回 name / introduction / official_website。
type SchoolLike = Partial<SchoolListItem> & {
  name?: string;
  school_code?: string;
  description?: string | null;
  introduction?: string | null;
  official_website?: string | null;
};

type ScoreLineLike = Partial<ScoreLine>;

// 后端布尔值可能是 boolean、数字或字符串，统一转换成真正的 boolean。
function toBoolean(value: unknown): boolean {
  return value === true || value === 1 || value === "1" || value === "true";
}

// 将院校列表 item 归一化，避免页面因为字段缺失或旧字段名而崩溃。
function normalizeSchoolItem(raw: SchoolLike): SchoolListItem {
  return {
    id: Number(raw.id ?? 0),
    school_name: String(raw.school_name ?? raw.name ?? ""),
    province: String(raw.province ?? ""),
    city: raw.city ?? null,
    school_type: raw.school_type ?? null,
    school_level: raw.school_level ?? null,
    is_985: toBoolean(raw.is_985),
    is_211: toBoolean(raw.is_211),
    is_double_first_class: toBoolean(raw.is_double_first_class),
    is_favorited: toBoolean((raw as SchoolLike & { is_favorited?: unknown }).is_favorited),
  };
}

// 院校详情在列表字段基础上补充 school_code 和 description。
function normalizeSchoolDetail(raw: SchoolLike): SchoolDetail {
  return {
    ...normalizeSchoolItem(raw),
    school_code: String(raw.school_code ?? ""),
    description: raw.description ?? raw.introduction ?? raw.official_website ?? null,
  };
}

// 分数线接口返回值也做一次兜底，保证页面始终拿到稳定字段类型。
function normalizeScoreLine(raw: ScoreLineLike): ScoreLine {
  return {
    year: Number(raw.year ?? 0),
    province: String(raw.province ?? ""),
    subject_type: String(raw.subject_type ?? ""),
    batch: String(raw.batch ?? ""),
    major_name: raw.major_name ?? null,
    min_score: Number(raw.min_score ?? 0),
    min_rank: raw.min_rank == null ? null : Number(raw.min_rank),
    avg_score: raw.avg_score == null ? null : Number(raw.avg_score),
    max_score: raw.max_score == null ? null : Number(raw.max_score),
  };
}

export async function searchSchools(
  params: SchoolSearchParams = {},
): Promise<SchoolSearchResponse> {
  // 查询流程：页面传入筛选条件 -> 调用后端 /schools/search -> normalize 后返回页面。
  const { data } = await apiClient.get<SchoolSearchResponse>("/schools/search", {
    params
  });

  // items 必须兜底成数组，否则页面读取 schools.length 时会报错。
  const items = Array.isArray(data?.items) ? data.items.map(normalizeSchoolItem) : [];
  const fallbackPageSize = params.page_size ?? (items.length > 0 ? items.length : 10);

  return {
    items,
    total: Number(data?.total ?? items.length),
    page: Number(data?.page ?? params.page ?? 1),
    page_size: Number(data?.page_size ?? fallbackPageSize),
  };
}

export async function getSchoolDetail(schoolId: number): Promise<SchoolDetail> {
  // 点击院校卡片后调用详情接口，右侧详情面板依赖这个结果渲染。
  const { data } = await apiClient.get<SchoolLike>(`/schools/${schoolId}`);
  return normalizeSchoolDetail(data ?? {});
}

export async function getSchoolScoreLines(
  schoolId: number,
  params: SchoolScoreLineParams = {},
): Promise<ScoreLine[]> {
  // 分数线参数都是可选的，页面可按省份、年份、科类、专业逐步过滤。
  const { data } = await apiClient.get<ScoreLineLike[]>(
    `/schools/${schoolId}/score-lines`,
    {
      params
    },
  );

  return Array.isArray(data) ? data.map(normalizeScoreLine) : [];
}

File: app/api/school.py
"""School search and detail APIs."""

from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from app.core.security import get_optional_current_user
from app.db.database import get_db
from app.models.school_schema import SchoolDetail, SchoolScoreLine, SchoolSearchResponse
from app.models.user import User
from app.services.activity_service import ActivityService
from app.services.favorite_service import FavoriteSchoolService
from app.services.school_service import SchoolService


router = APIRouter(prefix="/schools", tags=["schools"])


def clean_optional(value: str | None) -> str | None:
    """Trim optional string parameters and normalize blanks to None."""
    if value is None:
        return None
    stripped = value.strip()
    return stripped or None


@router.get("/search", response_model=SchoolSearchResponse)
def search_schools(
    school_name: str | None = Query(default=None, min_length=1, max_length=128),
    province: str | None = Query(default=None, min_length=1, max_length=32),
    school_level: str | None = Query(default=None, min_length=1, max_length=32),
    is_985: bool | None = Query(default=None),
    is_211: bool | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=100),
    db: Session = Depends(get_db),
) -> SchoolSearchResponse:
    """Search schools with optional filters."""
    try:
        total, items = SchoolService.search_schools(
            db,
            school_name=clean_optional(school_name),
            province=clean_optional(province),
            school_level=clean_optional(school_level),
            is_985=is_985,
            is_211=is_211,
            page=page,
            page_size=page_size,
        )
    except SQLAlchemyError as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to query schools. Please check MySQL connection and table initialization.",
        ) from exc

    return SchoolSearchResponse(total=total, page=page, page_size=page_size, items=items)


@router.get("/{school_id}", response_model=SchoolDetail)
def get_school_detail(
    school_id: int = Path(..., ge=1),
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_optional_current_user),
) -> SchoolDetail:
    """Return school detail by school ID."""
    try:
        school = SchoolService.get_school_detail(db, school_id)
    except SQLAlchemyError as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to query school detail. Please check MySQL connection and table initialization.",
        ) from exc

    if school is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"School {school_id} was not found.")

    response = SchoolDetail(
        **school,
        is_favorited=(
            FavoriteSchoolService.is_favorited(db, user=current_user, school_id=school_id)
            if current_user is not None
            else False
        ),
    )
    ActivityService.record_activity(
        db,
        user=current_user,
        activity_type="school_view",
        target_id=school_id,
        summary=response.school_name,
        payload={
            "school_id": school_id,
            "school_name": response.school_name,
            "province": response.province,
        },
    )
    return response


@router.get("/{school_id}/score-lines", response_model=list[SchoolScoreLine])
def get_school_score_lines(
    school_id: int = Path(..., ge=1),
    province: str | None = Query(default=None, min_length=1, max_length=32),
    year: int | None = Query(default=None, ge=2000, le=2100),
    subject_type: str | None = Query(default=None, min_length=1, max_length=32),
    major_name: str | None = Query(default=None, min_length=1, max_length=128),
    db: Session = Depends(get_db),
) -> list[SchoolScoreLine]:
    """Return historical score lines for one school."""
    try:
        if not SchoolService.school_exists(db, school_id):
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"School {school_id} was not found.")

        rows = SchoolService.get_school_score_lines(
            db,
            school_id=school_id,
            province=clean_optional(province),
            year=year,
            subject_type=clean_optional(subject_type),
            major_name=clean_optional(major_name),
        )
    except HTTPException:
        raise
    except SQLAlchemyError as exc:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to query score lines. Please check MySQL connection and table initialization.",
        ) from exc

    return [SchoolScoreLine(**row) for row in rows]
